### ISSUE TITLE
available qualifiers

### ISSUE THREAD

**Issue by rgetz** [analogdevicesinc/libiio#172]
The recent kernels are starting to support _available qualifiers, which is great -- this helps alot in figuring out what input specs and range various attributes have, but there are no existing helper functions in libiio to manage these, so each application is doing it by themselves.

root@analog:~# iio_attr -d ad9361-phy *available
dev 'ad9361-phy', attr 'dcxo_tune_fine_available', value :'[0 1 8191]'
dev 'ad9361-phy', attr 'xo_correction_available', value :'[40000000 1 40000000]'
dev 'ad9361-phy', attr 'dcxo_tune_coarse_available', value :'[0 1 63]'

which would be super easy to use something like:
iio_device_attr_available_longlong (dev, attr, &min, &step, &max);
iio_device_attr_available_double(dev, attr, &min, &step, &max);

but we also have:
dev 'ad9361-phy', attr 'calib_mode_available', value :'auto manual manual_tx_quad tx_quad rf_dc_offs rssi_gain_step'
dev 'ad9361-phy', attr 'ensm_mode_available', value :'sleep wait alert fdd pinctrl pinctrl_fdd_indep'
dev 'ad9361-phy', attr 'trx_rate_governor_available', value :'nominal highest_osr'

Which would be great to automatically populate pull downs (Which I think we already do)
I was thinking, something like:

iio_device_attr_available(dev, attr, &num, **array_of_strings);

the last one, I was not sure of - I thought MATLAB Mex had issues with arrays of strings? (@tfcollins - do you remember? or @mfornero ?)

Thanks
-Robin

**Comment by tfcollins** [comment]
Should handle it: https://www.mathworks.com/help/matlab/matlab_external/passing-arguments-to-shared-library-functions.html#f44412

Looks like an array of strings translates to a cell array of character vectors

**Comment by rgetz** [comment]
unless @larsclausen or @pcercuei have any comments - I will do something this weekend based on the above, and send a pull request.

-Robin

**Comment by tfcollins** [comment]
Add me as a reviewer and I can see if calllib and mex can interpret iio_device_attr_available output.

**Comment by pcercuei** [comment]
So all attributes have their corresponding _available now?
Or just some of them?

As for the implementation, I don't like the "array of strings" idea, especially in that case, because the strings would be dynamically allocated. Then is it up to the user to free every single one of them? I'd prefer the usual get_count / get_at_index method we already use all around in the API.

**Comment by rgetz** [comment]
The number of _availible atttibutes depends on the driver and the attribute type. Most of the AD9361 driver attributes supports _available now.

No problem on the implementation - that is what I will start with.

**Comment by pcercuei** [comment]
The "available" specifier for ints and doubles, is it a specification of the upstream iio? Or something that ADI came up with for the AD9361 driver?

**Comment by mhennerich** [comment]
https://static.lwn.net/kerneldoc/driver-api/iio/core.html
Look for IIO_AVAIL_RANGE.

struct iio_info
read_avail
    function to return the available values from the device. mask specifies which value. Note 0 means the available values for the channel in question. Return value specifies if a IIO_AVAIL_LIST or a IIO_AVAIL_RANGE is returned in vals. The type of the vals are returned in type and the number of vals is returned in length. For ranges, there are always three vals returned; min, step and max. For lists, all possible values are enumerated.

https://patchwork.kernel.org/patch/9391259/

**Comment by lclausen-adi** [comment]
You know that some driver will eventually change them at runtime.

**Comment by pcercuei** [comment]
It is non-volatile data, so you could populate it at context creation, that would be better I think.

**Comment by lclausen-adi** [comment]
There is no guarantee that it will not change. And that means there will be drivers where it changes.

**Comment by rgetz** [comment]
yes - I can see ranges or options changing, based on other attribute settings (like clocks for example) that might be exposed via run time controls.

run time updates it is.

**Comment by mhennerich** [comment]
We should not assume available attributes as constants
